homeassistant/components/cast/__init__.py

```diff
@@ -49,9 +58,58 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Cast from a config entry."""
     await home_assistant_cast.async_setup_ha_cast(hass, entry)
     hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    hass.data[DOMAIN] = {}
+    await async_process_integration_platforms(hass, DOMAIN, _register_cast_platform)
     return True
 
 
+class CastProtocol(Protocol):
+    """Define the format of cast platforms."""
+
+    async def async_get_media_browser_root_object(
+        self, content_filter: Callable[[BrowseMedia], bool]
+    ) -> list[BrowseMedia]:
+        """Create a list of root objects for media browsing."""
+
+    async def async_browse_media(
+        self,
+        hass: HomeAssistant,
+        media_content_type: str,
+        media_content_id: str,
+        content_filter: Callable[[BrowseMedia], bool],
+    ) -> BrowseMedia | None:
+        """Browse media.
+
+        Return a BrowseMedia object or None if the media does not belong to this platform.
+        """
+
+    async def async_play_media(
+        self,
+        hass: HomeAssistant,
+        cast_entity_id: str,
+        chromecast: Chromecast,
+        media_type: str,
+        media_id: str,
+    ) -> bool:
+        """Play media.
+
+        Return True if the media is played by the platform, False if not.
+        """
+
+
+async def _register_cast_platform(
+    hass: HomeAssistant, integration_domain: str, platform: CastProtocol
+):
+    """Register a cast platform."""
+    if (
+        not getattr(platform, "async_get_media_browser_root_object")
+        or not getattr(platform, "async_browse_media")
+        or not getattr(platform, "async_play_media")
+    ):
+        return
```

Comment:
Should we raise an error if the platform has an incorrect definition?

Reply:
Because we introduce this platform later, we can’t forbid other integrations to have a `cast` platform I think?

Reply:
I don't think we should allow any integration to create a platform with a name of another core integration, unless it's supported by the main integration. That namespace is already taken by the main integration.